project/sbt-changeling/Changeling.scala

```diff
@@ -0,0 +1,280 @@
+   * @param relaseNotes
+   *   The file into which the release notes will be written
+   * @returns
+   *   The TrelaseNotes
+   */
+  def renderReleaseNotes(
+      changeKinds: Seq[String],
+      version: String,
+      unreleasedDir: File,
+      releaseNotes: File): File = {
+
+    val changeDirOrder: File => Int = { d =>
+      changeKinds.indexOf(deNormalizeFileName(d.base.toString))
+    }
+
+    // The heading for the release notes
+    val versionHeading = s"## ${version} - ${LocalDate.now}"
+    // To render the directory structure described above into the required flatfile
+    // we just have to flatten the directory tree with the following transformations
+    //
+    // - transforming the change directory name into an H3 heading
+    // - get the contnet of each file in the change directory, and make it a
+    //   bullet list item
+    val changeSections = IO
+      .listFiles(unreleasedDir)
+      .sortBy(changeDirOrder)
+      .flatMap { changeDir =>
+        IO.listFiles(changeDir) match {
+          case Array() => Seq() // Omit sections with no entries
+          case changeEntries => {
+            val heading = s"### ${deNormalizeFileName(changeDir.base.toString)}"
+            val changes = changeEntries.map(readFile.andThen(mdBulletItem))
+            // We add an empty line between the heading and the changes and after the changes
+            (heading +: "" +: changes) ++ Seq("")
+          }
+        }
+      }
+    val notes = (versionHeading +: "" +: changeSections).mkString("\n")
+    IO.write(releaseNotes, notes)
+    releaseNotes
+  }
+
+  /**
+   * Prefix the content of `newFile` to the content of `target` and write the result to `target`
+   */
+  def concatFiles(front: File, target: File): Unit = {
+    val content = s"${IO.read(front)}\n${IO.read(target)}"
+    IO.write(target, content)
+  }
+
+  /**
+   * Move all the contents of the cange directories in `unrelasedDir` into a "trash" dir in `resourceDir`
+   *
+   * @return
+   *   the trash directory crated
+   */
+  def cleanChanges(resourceDir: File, unreleasedDir: File): Unit = {
+    // So we can recover dirs if needed
+    val trashDir = resourceDir / "changeling-trash" / LocalDate.now.toString
+    IO.createDirectory(trashDir)
+    val dirPaths = IO.listFiles(unreleasedDir)
+    val fileMoves = dirPaths.map(d => (d -> trashDir / d.base))
+    IO.move(fileMoves)
+    // Recreate the directories
+    IO.createDirectories(dirPaths)
+  }
+
+  private def mdBulletItem(s: String): String = {
+    val lineBreaks = "(\\r\\n?|\\n)"
+    val linesBreaksRemoved = s.replaceAll(lineBreaks, " ").trim()
+    s"- ${linesBreaksRemoved}"
+  }
+
+  // Heading -> Filename
+  private val normalizeFileName: String => String = _.replace(' ', '-').toLowerCase()
+  // Filename -> Heading
+  private val deNormalizeFileName: String => String = _.replace('-', ' ').capitalize
+  private val readFile: File => String = IO.read(_)
```

Comment:
You're calling `IO.read()` directly in all but one place; maybe this can go?

Reply:
Yeah, it’s probably not needed. I like to enable the function chaining in the one place it’s used, but that’s not really a huge readability win. I’ll ditch this. Thanks.